### PR TITLE
Clean log probs

### PR DESCRIPTION
**Pull request by shanmugamr1992**
# What does this PR do ?
1. Fixes skip_prompt_log_probs to make sure it works 
2. Made calculate log probs a little more efficient if materialize last token logits is set to true or is decode only
3. Added test case to check log probs properly

### CODE REVIEW

**Comment by kvareddy** [comment]
@santhnm2 @tdene  can you please take a look at this MR?

**Comment by shanmugamr1992** [comment]
/ok to test [0e5e279](https://github.com/NVIDIA/Megatron-LM/pull/2404/commits/0e5e2793142076303cfe046059164a5c613db1b9)

**Comment by shanmugamr1992** [comment]
/ok to test [97dcd9a](https://github.com/NVIDIA/Megatron-LM/pull/2404/commits/97dcd9a8cd8d129c4cadb68c3a3c4fda25a936d7)

**Comment by shanmugamr1992** [comment]
/ok to test [a07978d](https://github.com/NVIDIA/Megatron-LM/pull/2404/commits/a07978d290335763bfe810a5fcba59abbff4dd6e)

**Comment by shanmugamr1992** [comment]
/ok to test [a07978d](https://github.com/NVIDIA/Megatron-LM/pull/2404/commits/a07978d290335763bfe810a5fcba59abbff4dd6e)

**Comment by shanmugamr1992** [comment]
/ok to test [da4c0e8](https://github.com/NVIDIA/Megatron-LM/pull/2404/commits/da4c0e8d61280cca6c684e626776421e941bfb08)

**Review by tdene — APPROVED** [review]

**Comment by shanmugamr1992** [comment]
/ok to test [f94ae26](https://github.com/NVIDIA/Megatron-LM/pull/2404/commits/f94ae26bdd6f4f579d3811397da9cef88b8ed2b3)

**Comment by shanmugamr1992** [comment]
/ok to test [6fcbb84](https://github.com/NVIDIA/Megatron-LM/pull/2404/commits/6fcbb847de1ec73b9d2255e344f53b1a9486daa1)

File: tests/unit_tests/inference/engines/test_dynamic_engine.py [hunk 1]
```diff
@@ -989,14 +1159,14 @@ def test_chunked_prefill(self, model_provider: str):
     @pytest.mark.skipif(
         not is_fa_min_version("2.7.3"), reason="need latest flash attn for dynamic batching"
     )
-    @pytest.mark.parametrize("return_prompt_top_n_logprobs", [True, False])
+    @pytest.mark.parametrize("skip_prompt_log_probs", [True, False])
     @torch.inference_mode()
```
Comment (santhnm2) [inline tests/unit_tests/inference/engines/test_dynamic_engine.py:1223]: Can you mark this test as internal?

File: megatron/core/inference/engines/dynamic_engine.py [hunk 2]
```diff
@@ -850,22 +826,51 @@ def post_process_requests(
...
+            if request_log_probs is not None:
+                # Initialize lists if they don't exist
+                if not request.prompt_log_probs:
+                    request.prompt_log_probs = []
+                if not request.generated_log_probs:
+                    request.generated_log_probs = []
+
+                # For chunked prefill with materialize_only_last_token_logits, discard intermediate log probs
+                if (
+                    request_id == self.context.chunked_prefill_request_id
```
Comment (santhnm2) [inline megatron/core/inference/engines/dynamic_engine.py:841]: Do the tests also cover the chunked prefill case? If not can you add a test for this?

File: tests/unit_tests/inference/engines/test_dynamic_engine.py (hunk 1 above)
Comment (shanmugamr1992) [reply tests/unit_tests/inference/engines/test_dynamic_engine.py:1223]: It is already marked internal

**Comment by shanmugamr1992** [comment]
/ok to test [62fdc5a](https://github.com/NVIDIA/Megatron-LM/pull/2404/commits/62fdc5ae4508910133632f5e2c803cc13be07214)

**Comment by shanmugamr1992** [comment]
/ok to test [2ce454e](https://github.com/NVIDIA/Megatron-LM/pull/2404/commits/2ce454e27834fbcb947102e2a41241e03a28d52b)

**Review by yaoyu-33 — APPROVED** [review]

File: megatron/core/inference/engines/dynamic_engine.py (hunk 2 above)
Comment (shanmugamr1992) [reply megatron/core/inference/engines/dynamic_engine.py:841]: added
